lib/common/redis.c

```diff
@@ -0,0 +1,219 @@
+            conn->cb.on_connection_failure(redis->errstr, conn->cb_data);
+        }
+    }
+}
+
+h2o_redis_conn_t *h2o_redis_connect(h2o_loop_t *loop, const char *host, uint16_t port, h2o_redis_callbacks_t cb, void *cb_data)
+{
+    h2o_redis_conn_t *conn = h2o_mem_alloc(sizeof(*conn));
+    *conn = (h2o_redis_conn_t){NULL};
+
+    redisAsyncContext *redis = redisAsyncConnect(host, port);
+
+    if (redis == NULL || attach_loop(redis, loop) != 0) {
+        /* failed to allocate memory */
+        goto Error;
+    }
+
+    conn->cb = cb;
+    conn->cb_data = cb_data;
+
+    if (redis->err == REDIS_OK) {
+        conn->redis = redis;
+        conn->redis->data = conn;
+        redisAsyncSetConnectCallback(redis, on_redis_connect);
+        redisAsyncSetDisconnectCallback(redis, on_redis_disconnect);
+    } else {
+        /* some connection failures can be detected at this time */
+        if (conn->cb.on_connection_failure != NULL) {
+            conn->cb.on_connection_failure(redis->errstr, conn->cb_data);
+        }
+        redisAsyncFree(redis);
+    }
+
+    return conn;
+
+Error:
+    if (redis != NULL)
+        redisAsyncFree(redis);
+    free(conn);
+    return NULL;
+}
+
+void h2o_redis_free(h2o_redis_conn_t *conn)
+{
+    if (conn->redis != NULL) {
+        redisAsyncDisconnect(conn->redis);
+    }
+    free(conn);
+}
+
+/* redis socket adapter */
+
+struct st_redis_socket_data {
+    redisAsyncContext *context;
+    h2o_socket_t *socket;
+};
+
+static void on_read(h2o_socket_t* sock, const char *err)
+{
+    struct st_redis_socket_data *p = (struct st_redis_socket_data *)sock->data;
+    redisAsyncHandleRead(p->context);
+}
+
+static void on_write(h2o_socket_t *sock, const char *err)
+{
+    struct st_redis_socket_data *p = (struct st_redis_socket_data *)sock->data;
+    redisAsyncHandleWrite(p->context);
+}
+
+static void socket_add_read(void *privdata) {
+    struct st_redis_socket_data *p = (struct st_redis_socket_data *)privdata;
+    h2o_socket_read_start(p->socket, on_read);
+}
+
+
+static void socket_del_read(void *privdata) {
+    struct st_redis_socket_data *p = (struct st_redis_socket_data *)privdata;
+    h2o_socket_read_stop(p->socket);
+}
+
+
+static void socket_add_write(void *privdata) {
+    struct st_redis_socket_data *p = (struct st_redis_socket_data *)privdata;
+    if (! h2o_socket_is_writing(p->socket)) {
+        h2o_socket_notify_write(p->socket, on_write);
+    }
+}
+
+static void socket_cleanup(void *privdata) {
+    struct st_redis_socket_data *p = (struct st_redis_socket_data *)privdata;
+    h2o_socket_close(p->socket);
+    p->context->c.fd = -1; /* prevent hiredis from closing fd twice */
+    free(p);
+}
+
+static int attach_loop(redisAsyncContext* ac, h2o_loop_t* loop) {
+    redisContext *c = &(ac->c);
+
+    struct st_redis_socket_data *p = (struct st_redis_socket_data *)malloc(sizeof(*p));
+    if (p == NULL) {
+        return -1;
+    }
+    memset(p, 0, sizeof(*p));
```

Comment:
As a general rule, I recommend using compound assignment for initializing a structure. In this case, `*p = (struct ...){NULL}`?